File: app/src/main/java/com/example/recipeclipper/data/LibraryPolicy.kt
package com.example.recipeclipper.data

import com.example.recipeclipper.data.flags.FeatureFlags
import com.example.recipeclipper.data.flags.Flag
import com.example.recipeclipper.data.model.LibraryLimit
import kotlinx.coroutines.flow.Flow
import kotlinx.coroutines.flow.combine
import kotlinx.coroutines.flow.distinctUntilChanged
import kotlinx.coroutines.flow.flowOf
import javax.inject.Inject
import javax.inject.Singleton

/**
 * Which [LibraryLimit] applies (#107): the `freeTier` flag, then the unlock, bought or the
 * Developer settings override. An interface so repository tests can fix one.
 */
interface LibraryPolicy {
    val limit: LibraryLimit
    val limits: Flow<LibraryLimit>

    /** The app as it was before #107, for tests and callers with no policy in mind. */
    object HistoryOnly : LibraryPolicy {
        override val limit: LibraryLimit = LibraryLimit.History(LibraryLimit.HISTORY_RECIPES)
        override val limits: Flow<LibraryLimit> = flowOf(limit)
    }
}

@Singleton
class DefaultLibraryPolicy @Inject constructor(
    private val flags: FeatureFlags,
    private val entitlements: Entitlements
) : LibraryPolicy {

    override val limit: LibraryLimit
        get() = LibraryLimit.of(
            freeTier = flags.isOn(Flag.FREE_TIER),
            unlocked = entitlements.state.value.unlocked || flags.unlockedOverride
        )

    override val limits: Flow<LibraryLimit> =
        combine(flags.values, flags.unlockedOverrides, entitlements.state) { values, override, state ->
            LibraryLimit.of(values.isOn(Flag.FREE_TIER), state.unlocked || override)
        }.distinctUntilChanged()
}

File: ios/RecipeClipper/App/LibraryPolicy.swift
import Foundation
import Observation

/// Which `LibraryLimit` applies (#107): the `freeTier` flag, then the unlock, bought or the
/// Developer settings override. Android's `DefaultLibraryPolicy`.
///
/// The repository reads the limit from `mirror` (the App Group suite), not from here, because
/// the share extension saves recipes in its own process and never sees the flags or StoreKit.
/// So this keeps the mirror current: at start, and on every change it observes.
@MainActor
@Observable
final class LibraryPolicy {
    @ObservationIgnored private let flags: FeatureFlags
    @ObservationIgnored private let entitlements: Entitlements
    @ObservationIgnored private let mirror: DefaultsLibraryLimit?

    init(flags: FeatureFlags, entitlements: Entitlements, mirror: DefaultsLibraryLimit? = nil) {
        self.flags = flags
        self.entitlements = entitlements
        self.mirror = mirror
    }

    var limit: LibraryLimit {
        LibraryLimit.of(
            freeTier: flags.isOn(.freeTier),
            unlocked: entitlements.state.unlocked || flags.unlockedOverride
        )
    }

    /// Writes the limit to the mirror now and after every change, for as long as the app runs.
    func startMirroring() {
        guard let mirror else { return }
        let current = withObservationTracking { limit } onChange: { [weak self] in
            Task { @MainActor in self?.startMirroring() }
        }
        mirror.store(current)
    }
}
